src/libcephfs.cc

```diff
@@ -2019,31 +2019,114 @@ extern "C" int64_t ceph_ll_writev(class ceph_mount_info *cmount,
   return (cmount->get_client()->ll_writev(fh, iov, iovcnt, off));
 }
 
+struct ceph_ll_readv_writev_buffer {
+  ceph_ll_readv_writev_buffer() : iov() {};
+  ~ceph_ll_readv_writev_buffer() {
+    delete iov;
+  }
+  void prepare_iovs() {
+    ceph_assert(bl.get_num_buffers() <= IOV_MAX);
+
+    iovcnt = bl.get_num_buffers();
+    iov = new struct iovec [iovcnt];
+    auto pb = std::cbegin(bl.buffers());
+    for (auto n = 0; n < iovcnt; n++) {
```

Comment:
if it is zerocpy and iovcnt is > 1 then this iteration would go through, do you think we should add some condition to check if zerocpy is true then make sure iovcnt is 1?

Reply:
We're fine if iovcnt is larger than 1 as long as it's <= IOV_MAX

Reply:
if let's say iovcnt is 5, it's zerocopy so it should've had 1, now it passes the `ceph_assert(bl.get_num_buffers() <= IOV_MAX);` check, the iteration would go on but it shouldn't since zerocopy expects only 1 as iovcnt, right?

Reply:
Ah, I see the source of confusion.

For read, we pass in an iovec of length 1, with the entry having the iov_len being the read size and iov_base is NULL. The caller expects that iovec to be replaced with a different iovec from the file system, so the fact that the bufferlist has 5 buffers in it is fine, cephfs creates an iovec of length 5 with those buffers and returns that.

Reply:
Note the new on line 2031.

Reply:
but what if caller made an error and passed iovcnt of length 5 instead of 1? This code assumes that the iovcnt passed is always correct but it might not be true always. So I was proposing a check like:

```
if (io_info->zerocopy) {
ceph_assert (iovcnt == 1);
}
```

Reply:
Oh, yea, that might make sense

Reply:
OK, I just updated and removed the DRAFT.

Please review